crates/neon/src/types_impl/buffer/mod.rs

```diff
@@ -0,0 +1,281 @@
+        &mut self.data
+    }
+}
+
+impl<'a, T> Drop for Ref<'a, T> {
+    fn drop(&mut self) {
+        let mut ledger = self.ledger.borrow_mut();
+        let range = Ledger::slice_to_range(self.data);
+        let i = ledger.shared.iter().rposition(|r| r == &range).unwrap();
+
+        ledger.shared.remove(i);
+    }
+}
+
+impl<'a, T> Drop for RefMut<'a, T> {
+    fn drop(&mut self) {
+        let mut ledger = self.ledger.borrow_mut();
+        let range = Ledger::slice_to_range(self.data);
+        let i = ledger.owned.iter().rposition(|r| r == &range).unwrap();
+
+        ledger.owned.remove(i);
+    }
+}
+
+#[derive(Eq, PartialEq)]
+/// An error returned by [`TypedArray::try_borrow`] or [`TypedArray::try_borrow_mut`] indicating
+/// that a mutable borrow would overlap with another borrow.
+///
+/// [`BorrowError`] may be converted to an exception with [`ResultExt::or_throw`].
+pub struct BorrowError {
+    _private: (),
+}
+
+impl BorrowError {
+    fn new() -> Self {
+        BorrowError { _private: () }
+    }
+}
+
+impl Error for BorrowError {}
+
+impl Display for BorrowError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt("Borrow overlaps with an active mutable borrow", f)
+    }
+}
+
+impl Debug for BorrowError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("BorrowError").finish()
+    }
+}
+
+impl<T> ResultExt<T> for Result<T, BorrowError> {
+    fn or_throw<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<T> {
+        self.or_else(|_| cx.throw_error("BorrowError"))
+    }
+}
+
+/// Represents a typed region of an [`ArrayBuffer`](crate::types::JsArrayBuffer).
+///
+/// A `Region` can be created via the
+/// [`Handle<JsArrayBuffer>::region()`](crate::handle::Handle::region) or
+/// [`JsTypedArray::region()`](crate::types::JsTypedArray::region) methods.
+///
+/// A region is **not** checked for validity until it is converted
+/// a typed array via [`to_typed_array()`](Region::to_typed_array) or
+/// [`JsTypedArray::from_region()`](crate::types::JsTypedArray::from_region).
+///
+/// # Example
+///
+/// ```
+/// # use neon::prelude::*;
+/// # fn f(mut cx: FunctionContext) -> JsResult<JsUint32Array> {
+/// // Allocate a 16-byte ArrayBuffer and a uint32 array of length 2 (i.e., 8 bytes)
+/// // starting at byte offset 4 of the buffer:
+/// //
+/// //       0       4       8       12      16
+/// //      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// // buf: | | | | | | | | | | | | | | | | |
+/// //      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// //               ^       ^
+/// //               |       |
+/// //              +-------+-------+
+/// //         arr: |       |       |
+/// //              +-------+-------+
+/// //               0       1       2
+/// let buf = cx.array_buffer(16)?;
+/// let arr = JsUint32Array::from_region(&mut cx, buf.region(4, 2))?;
+/// # Ok(arr)
+/// # }
+/// ```
+#[derive(Clone, Copy)]
```

Comment:
I'm not sure `Region` should be `Copy`. It's a little large, but on the line.

Reply:
I've left it as `Copy` (which you mentioned on Slack is your preference) but changed all the methods to take `&Region`. I'm persuadable either way on `Copy`.